MAINT: optimize: expose `NonConvergence`

j-bowhay opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
closes #20100 
#### What does this implement/fix?
<!--Please explain your changes.-->
Currently, this expectation is only exposed through the deprecated namespace `scipy.optimize.nonlin`. This PR exposes it in the main `scipy.optimize` namespace so that it can be accessed without a deprecation warning.
#### Additional information
<!--Any additional information you think is important.-->

j-bowhay committed:
MAINT: optimize: expose `NonConvergence`

rgommers reviewed (approved):
Thanks @j-bowhay. Changes LGTM. In addition, it would be useful to have a test that uses `optimize.NoConvergence` to ensure in the test suite that that exception now exists.

j-bowhay commented:
Thanks for the review @rgommers, while adding the test it looks like something isn't quite right here as when `root` calls `nonlin_solve` it sets `raise_exception=False` so I am not sure it is possible to raise this exception through the public API. @physics-enthusiast what use case do you need access to `NoConvergence`.

Either way the docs for `root` need updating as this statement is incorrect:
![image](https://github.com/scipy/scipy/assets/60778417/fa09bf85-0a30-45aa-8bc0-dfdc233f4164)

physics-enthusiast commented:
The [named wrapper functions](https://github.com/scipy/scipy/blob/74ab3614d9d94004790bd49df1ad0cba4d34d34a/scipy/optimize/_nonlin.py#L1541) do not specify ```raise_exception```, which should cause it to default to ```True```. For example, I can call ```scipy.optimize.newton_krylov``` as opposed to ```scipy.optimize.root(method='krylov')```.

j-bowhay committed:
DOC: move location of NoConvergence in toc tree

j-bowhay committed:
DOC: remove incorrect statement about NoConvergence error

j-bowhay committed:
TST: add test with NoConvergence

j-bowhay commented:
Thanks for the explanation @physics-enthusiast 

@rgommers I have updated the PR with a test, relocated the entry in the toc tree, and fixed the incorrect documentation.

rgommers commented:
Hmm, Chocolatey has decided that it wants credentials or some such thing. Guess we'll ignore that here for now.

rgommers reviewed (approved):
Let's give this a go. Thanks @j-bowhay and @physics-enthusiast
